root/hist/hist/FitGradBenchmarks.cxx

```diff
@@ -0,0 +1,259 @@
+      for (int i = 0; i <3; ++i)
+        stepSizes[i] = 0.01*std::abs(fitFunc->GetParameter(i));
+
+      // set gaussians parameters
+      for (int i = 0; i < ng; ++i) {
+         fitFunc->SetParameter(TString::Format("A%d", i), a[i]);
+         fitFunc->SetParameter(TString::Format("mu%d", i), mu[i]);
+         fitFunc->SetParameter(TString::Format("sig%d", i), sig[i]);
+         stepSizes[3 + 3*i] = 0.01*std::abs(a[i]);
+         stepSizes[3 + 3*i + 1] = h1->GetBinWidth(1);
+         stepSizes[3 + 3*i + 2] = 0.1*h1->GetBinWidth(1);
+      }
+
+      if (debug>1)
+         fitFunc->Print("V");
+
+      // fitFunc->Draw();
+      if (debug>1)
+         fitFunc->SetNpx(5 * nbins);
+
+      if (useCLAD)
+         fitFunc->GetFormula()->GenerateGradientPar();
+
+      initParams = std::vector<double>(fitFunc->GetParameters(), fitFunc->GetParameters() + fitFunc->GetNpar());
+
+      // avoid computing Hessian at end of minimization
+      ROOT::Math::MinimizerOptions::SetDefaultStrategy(0);
+      // try using Minuit whenever possible
+      ROOT::Math::MinimizerOptions::SetDefaultMinimizer("Minuit2");
+   }
+
+   bool Fit()
+   {
+      TString fitOpt = "S  0 ";
+      if (debug<2)
+         fitOpt += TString(" Q ");
+      else
+         fitOpt += TString(" V ");
+      if (useCLAD)
+         fitOpt += TString(" G");
+      // need to set initial parameters and step size to make all fits equivalent
+      // when using AD is quite important having good step sizes for having a good MnSeed
+      fitFunc->SetParameters(initParams.data());
+      std::vector<double> emptyStepSizes(fitFunc->GetNpar());
+      // need to set step size to have fits
+      fitFunc->SetParErrors(emptyStepSizes.data());
+      TString type = (useCLAD) ? "Clad" : "numerical";
+
+      TStopwatch w;
+      if (debug)  w.Start();
+      auto res = h1->Fit(fitFunc, fitOpt);
+
+    if (debug) {
+      std::cout << "Using " << type << " differentiations : ncalls " << res->NCalls() << " elapsed time : ";
+      w.Print();
+    }
+
+
+    if (debug>1)
+         res->Print();
+
+      bool ok = (res->Status() == 0);
+      if (!ok)
+         Error("ManyGausFit", "Error fitting with many gaussians, return status is %d", res->Status());
+
+      return ok;
+   }
+
+   ~ManyGausFit()
+   {
+      delete fitFunc;
+      delete h1;
+   }
+};
+// default values
+int ngaussians = 30;
+int nbins = 1000;   // nbins should be at least 10* ngaussians
+int printLevel = 0;
+
+static void BM_TFormula_ManyGausFit_Clad(benchmark::State &state)
+{
+   ManyGausFit test(ngaussians, nbins, true,  printLevel);
+   for (auto _ : state) {
+      test.Fit();
+   }
+}
+
+BENCHMARK(BM_TFormula_ManyGausFit_Clad);
+
+static void BM_TFormula_ManyGausFit_Numerical(benchmark::State &state)
+{
+   ManyGausFit test(ngaussians, nbins, false, printLevel);
+   for (auto _ : state) {
+      test.Fit();
+   }
+}
+
+BENCHMARK(BM_TFormula_ManyGausFit_Numerical);
+
+// Define our main.
+//BENCHMARK_MAIN();
```

Comment:
Why this is not sufficient?

Reply:
I define a main function in order to parse some command line parameter to control the test, for example increasing the number of parameters to fit

Reply:
I see. It may be beneficial to ask rootbench to automatically make a scalability comparison:
```
diff --git a/root/hist/hist/FitGradBenchmarks.cxx b/root/hist/hist/FitGradBenchmarks.cxx
index 476d29e..057d0fe 100644
--- a/root/hist/hist/FitGradBenchmarks.cxx
+++ b/root/hist/hist/FitGradBenchmarks.cxx
@@ -27,7 +27,7 @@ static void BM_TFormula_GausFit_Clad(benchmark::State &state) {
   delete h1;
 }
 
-BENCHMARK(BM_TFormula_GausFit_Clad);
+BENCHMARK(BM_TFormula_GausFit_Clad)->Unit(benchmark::kMicrosecond);
 
 static void BM_TFormula_GausFit_Numerical(benchmark::State &state) {
   auto f1 = new TF1("f1", "gaus");
@@ -46,7 +46,7 @@ static void BM_TFormula_GausFit_Numerical(benchmark::State &state) {
   delete h1;
 }
 
-BENCHMARK(BM_TFormula_GausFit_Numerical);
+BENCHMARK(BM_TFormula_GausFit_Numerical)->Unit(benchmark::kMicrosecond);
 
 // test class for building fit with many parameters (many gaussian + polynomial)
 class ManyGausFit {
@@ -202,29 +202,33 @@ public:
    }
 };
 // default values
-int ngaussians = 30;
-int nbins = 1000;   // nbins should be at least 10* ngaussians
 int printLevel = 0;
 
 static void BM_TFormula_ManyGausFit_Clad(benchmark::State &state)
 {
+   unsigned ngaussians = state.range(0);
+   unsigned nbins = 100 * ngaussians;
    ManyGausFit test(ngaussians, nbins, true,  printLevel);
    for (auto _ : state) {
       test.Fit();
    }
 }
 
-BENCHMARK(BM_TFormula_ManyGausFit_Clad);
+BENCHMARK(BM_TFormula_ManyGausFit_Clad)->Unit(benchmark::kMicrosecond)
+->DenseRange(8, 64, 8);
 
 static void BM_TFormula_ManyGausFit_Numerical(benchmark::State &state)
 {
+   unsigned ngaussians = state.range(0);
+   unsigned nbins = 100 * ngaussians;
    ManyGausFit test(ngaussians, nbins, false, printLevel);
    for (auto _ : state) {
       test.Fit();
    }
 }
 
-BENCHMARK(BM_TFormula_ManyGausFit_Numerical);
+BENCHMARK(BM_TFormula_ManyGausFit_Numerical)->Unit(benchmark::kMicrosecond)
+->DenseRange(8, 64, 8);
 
 // Define our main.
 //BENCHMARK_MAIN();
@@ -242,14 +246,6 @@ int main(int argc, char **argv) {
         std::cout << "---running in very verbose mode" << std::endl;
         printLevel = 2;
       }
-      else if (arg == "-nfunc") {
-         std::cout << "--- running using " << ngaussians << " functions" << std::endl;
-         ngaussians = atoi(argv[++i]);
-      }
-       else if (arg == "-nbins") {
-         std::cout << "--- running using " << nbins << " bins" << std::endl;
-         nbins = atoi(argv[++i]);
-      }
    }
 
 ```
 
Going to 64 or 128 makes the benchmark too slow to run on the continuous performance infrastructure but going to 24 seems reasonable.

A strange effect can be seen at /56 and /72 where clad shows timing better than at /40 and around /48 respectively. 

I am a little confused, as the diff I propose gives me:

```
./root/hist/hist/FitGradBenchmarks
2020-12-19 19:59:38
Running ./root/hist/hist/FitGradBenchmarks
Run on (8 X 8300 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
Load Average: 2.74, 2.87, 3.25
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations
-------------------------------------------------------------------------------
BM_TFormula_GausFit_Clad                   4050 us         4050 us          166
BM_TFormula_GausFit_Numerical              2357 us         2357 us          262
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Clad/8           426782 us       426780 us            2
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Clad/16         4714236 us      4714224 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Clad/24        19373080 us     19373052 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Clad/32        50276601 us     50276536 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Clad/40       106593384 us    106593169 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Clad/48       199842059 us    199839571 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Clad/56       336228844 us    336223681 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Clad/64       540560718 us    540555134 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Numerical/8      143316 us       143312 us            5
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Numerical/16    1078621 us      1078599 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Numerical/24    3426112 us      3426108 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Numerical/32    7983433 us      7983428 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Numerical/40   15551319 us     15551313 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Numerical/48   27331356 us     27331346 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Numerical/56   43774259 us     43774240 us            1
Warning in <ROOT::Math::FitConfig::CreateMinimizer>: Could not create the Minuit2 minimizer. Try using the minimizer Minuit
BM_TFormula_ManyGausFit_Numerical/64   66523336 us     66523307 us 
```

Do you know what might be going wrong?

Reply:
Are you sure you are using the ROOT master where Clad can be enabled in fitting ? 
Otherwise clad it is not actually used and the numerical differentiation TF1 is used that is much slower that the one used in Minuit. 
I would suggest you also to enabled Minuit2 when building ROOT (configure with -Dminuit2=On). I will add soon a patch in Minuit2 that will improve significantly the performance of the fitting when using an external gradient (e.g. clad)

Reply:
Ok, that is better:
```
 ./root/hist/hist/FitGradBenchmarks
2020-12-21 18:42:05
Running ./root/hist/hist/FitGradBenchmarks
Run on (8 X 8300 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
Load Average: 2.99, 3.66, 4.90
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
-------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations
-------------------------------------------------------------------------------
BM_TFormula_GausFit_Clad                   2261 us         2261 us          321
BM_TFormula_GausFit_Numerical              2220 us         2220 us          277
BM_TFormula_ManyGausFit_Clad/8            29512 us        29511 us           24
BM_TFormula_ManyGausFit_Clad/16          174089 us       174086 us            4
BM_TFormula_ManyGausFit_Clad/24          475320 us       475318 us            2
BM_TFormula_ManyGausFit_Clad/32         1077862 us      1077855 us            1
BM_TFormula_ManyGausFit_Clad/40         1886996 us      1886981 us            1
BM_TFormula_ManyGausFit_Clad/48         2895262 us      2895216 us            1
BM_TFormula_ManyGausFit_Clad/56         4462770 us      4462737 us            1
BM_TFormula_ManyGausFit_Clad/64         6086814 us      6086761 us            1
BM_TFormula_ManyGausFit_Numerical/8      139701 us       139700 us            5
BM_TFormula_ManyGausFit_Numerical/16    1036581 us      1036575 us            1
BM_TFormula_ManyGausFit_Numerical/24    3298170 us      3298158 us            1
BM_TFormula_ManyGausFit_Numerical/32    7636679 us      7636651 us            1
BM_TFormula_ManyGausFit_Numerical/40   14895533 us     14895474 us            1
BM_TFormula_ManyGausFit_Numerical/48   26005987 us     26005869 us            1
BM_TFormula_ManyGausFit_Numerical/56   30612857 us     30612727 us            1
BM_TFormula_ManyGausFit_Numerical/64   61032731 us     61032485 us            1

```
I agree now the scalability looks very different. I like also the much flattened curve with `ngaussians`. Should we also test scalability wrt `nbins`, now with my patch `nbins = 100 * ngaussians`.

We need to make sure we benchmark against root configured with -Dminuit2=On. @oshadura, is that already the case?

Reply:
@efremale, fyi, we seem to have finally made it and we scale well. Well done!